service/history/workflowExecutionContext.go

```diff
@@ -1127,3 +1169,59 @@ func (c *workflowExecutionContextImpl) resetWorkflowExecution(
 func (c *workflowExecutionContextImpl) getQueryRegistry() QueryRegistry {
 	return c.queryRegistry
 }
+
+func (c *workflowExecutionContextImpl) reapplyEvents(
+	ctx context.Context,
+	events *persistence.WorkflowEvents,
+) error {
+
+	reapplyEvents := []*workflow.HistoryEvent{}
+	// TODO: need to implement Reapply policy
+	for _, event := range events.Events {
+		switch event.GetEventType() {
+		case workflow.EventTypeWorkflowExecutionSignaled:
+			reapplyEvents = append(reapplyEvents, event)
+		}
+	}
+	// there is no event to reapply
+	if len(reapplyEvents) == 0 {
+		return nil
+	}
+
+	domainID := events.DomainID
+	execution := &workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr(events.WorkflowID),
+	}
+	domainCache := c.shard.GetDomainCache()
+	clientBean := c.shard.GetService().GetClientBean()
+	serializer := c.shard.GetService().GetPayloadSerializer()
+	domainEntry, err := domainCache.GetDomainByID(domainID)
+	if err != nil {
+		return err
+	}
+
+	// The active cluster of the domain is the same as current cluster.
+	// Use the history from the same cluster to reapply events
+	reapplyEventsDataBlob, err := serializer.SerializeBatchEvents(reapplyEvents, common.EncodingTypeThriftRW)
+	if err != nil {
+		return err
+	}
+	request := &workflow.ReapplyEventsRequest{
+		DomainName:        common.StringPtr(domainEntry.GetInfo().Name),
+		WorkflowExecution: execution,
+		Events:            reapplyEventsDataBlob.ToThrift(),
+	}
+
+	activeCluster := domainEntry.GetReplicationConfig().ActiveClusterName
+	if activeCluster == c.shard.GetClusterMetadata().GetCurrentClusterName() {
+		return c.shard.GetEngine().ReapplyEvents(ctx, &history.ReapplyEventsRequest{
```

Comment:
history engine's ReapplyEvents API should take in events, not blob
history service's ReapplyEvents API should take in blob, not events

Reply:
done